src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -98,40 +106,123 @@ public bool TryUnparse(StringBuilder sb)
         }
 
+        public sealed class ColumnInfo
+        {
+            public readonly string Input;
+            public readonly string Output;
+            public readonly string WeightColumn;
+            public readonly int Rank;
+            public readonly int Oversampling;
+            public readonly bool Center;
+            public readonly int? Seed;
+
+            /// <summary>
+            /// Describes how the transformer handles one column pair.
+            /// </summary>
+            public ColumnInfo(string input,
+                              string output,
+                              string weightColumn = Defaults.WeightColumn,
+                              int rank = Defaults.Rank,
+                              int overSampling = Defaults.Oversampling,
+                              bool center = Defaults.Center,
+                              int? seed = null)
+            {
+                Input = input;
+                Output = output;
+                WeightColumn = weightColumn;
+                Rank = rank;
+                Oversampling = overSampling;
+                Center = center;
+                Seed = seed;
+            }
+
+            // The following functions and properties are all internal and used for simplifying the
+            // Transformer and Mapper code.
+
+            internal ColumnInfo((string input, string output) columnPair)
+            {
+                Input = columnPair.input;
+                Output = columnPair.output;
+            }
+
+            internal void SetSchema(Schema schema)
+            {
+                _schema = schema;
+            }
+
+            private Schema _schema;
+
+            internal int InputIndex
+            {
+                get
+                {
+                    Contracts.AssertValue(_schema);
+                    // Column names are already checked by PcaTransform
+                    _schema.TryGetColumnIndex(Input, out int index);
+                    return index;
+                }
+            }
+
+            internal ColumnType InputType
+            {
+                get
+                {
+                    Contracts.AssertValue(_schema);
+                    return _schema[Input].Type;
+                }
+            }
+
+            internal int WeightColumnIndex
+            {
+                get
+                {
+                    Contracts.AssertValue(_schema);
+                    var index = -1;
+                    if (WeightColumn != null)
+                    {
+                        if (!_schema.TryGetColumnIndex(WeightColumn, out index))
+                            throw Contracts.Except("Weight column '{0}' does not exist.", WeightColumn);
+                        Contracts.CheckUserArg(_schema[index].Type == NumberType.Float, nameof(WeightColumn));
+                    }
+                    return index;
+                }
+            }
+        }
+
         private sealed class TransformInfo
```

Comment:
>TransformInfo [](start = 29, length = 13)

I suggest folding `TransformInfo` and `ColumnInfo` in one class, if possible

Reply:
I had thought about it but it's too much work. The main thing is that TransformInfo get loaded/saved with ModelContextInfo and has a different life cycle.

---
In reply to: [227069182](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227069182) [](ancestors = 227069182)